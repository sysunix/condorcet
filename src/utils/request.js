import firebase, { db } from "../firebase";

export const createPoll = async poll => {
  const pollWithTimestap = {
    ...poll,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db.collection("polls").add(pollWithTimestap);
};

export const createVote = async (vote, pollId, userId) => {
  const voteWithTimestap = {
    ...vote,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db
    .collection("polls")
    .doc(pollId)
    .collection("votes")
    .doc(userId)
    .set(voteWithTimestap);
};
