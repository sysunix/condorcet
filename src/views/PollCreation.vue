<template>
  <form class="w-full pt-4" autocomplete="off" @keypress.enter.prevent>
    <div class="flex flex-wrap -mx-3 mb-6">
      <div class="w-full md:w-1/2 px-3 mb-6 md:mb-0">
        <label
          class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
          for="question"
          >{{ $t("poll.fields.question") }} *</label
        >
        <input
          id="question"
          v-model="question"
          data-testid="form-question"
          class="appearance-none block w-full bg-gray-200 text-gray-700 border rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"
          type="text"
        />

        <label
          class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
          for="description"
          >{{ $t("poll.fields.description") }}</label
        >
        <textarea
          v-model="description"
          data-testid="form-description"
          class="appearance-none block w-full bg-gray-200 text-gray-700 border rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"
        ></textarea>

        <div class="md:flex md:items-center mb-6">
          <label
            class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            for="description"
            >{{ $t("poll.fields.isPublic") }}</label
          >
          <input
            id="isPublic"
            v-model="isPublic"
            data-testid="form-isPublic"
            class="mr-2 leading-tight"
            type="checkbox"
            name="isPublic"
          />
          <label
            for="isPublic"
            class="md:w-2/3 block text-gray-500 font-bold inline-block"
          >
            <span class="text-sm">{{ isPublicText }}</span>
          </label>
        </div>
      </div>
      <div class="w-full md:w-1/2 px-3">
        <label
          class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
          for="answer"
          >{{ $t("poll.fields.answer") }} *</label
        >
        <div class="flex">
          <input
            id="answer"
            ref="answerInput"
            v-model="answer"
            data-testid="form-answer"
            class="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            type="text"
            @keypress.enter="addAnswer"
          />
          <button
            data-testid="form-addAnswer"
            class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            @click.prevent="addAnswer"
          >
            {{ $t("global.add") }}
          </button>
        </div>

        <ul class="list-disc mt-6 ml-10">
          <li
            v-for="answerProposal in answers"
            :key="answerProposal.slug"
            class="mt-2"
          >
            {{ answerProposal.value }}
            <button
              :data-testid="`form-answerProposal-${answerProposal.slug}`"
              class="bg-red-500 hover:bg-red-700 text-white font-bold w-6 h-6 rounded-full float-right"
              @click.prevent="removeAnswer(answerProposal.slug)"
            >
              -
            </button>
          </li>
        </ul>
      </div>

      <div class="w-full mt-4 px-3 flex justify-center">
        <button
          data-testid="form-submit"
          :disabled="disabledSubmit"
          class="w-1/3 bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
          :class="disabledSubmit && 'opacity-50 cursor-not-allowed'"
          @click.prevent="submitPoll"
        >
          {{ $t("global.create") }}
        </button>
      </div>
    </div>
  </form>
</template>

<script>
import { mapState, mapActions } from "vuex";
import slugify from "slugify";
import randomString from "random-string";
import { createPoll } from "../utils/request";

export default {
  name: "PollCreation",
  data() {
    return {
      question: "",
      description: "",
      answer: "",
      answers: [],
      isPublic: false
    };
  },
  computed: {
    ...mapState({
      userId: state => state.user.id
    }),
    isPublicText() {
      return this.isPublic ? "Oui" : "Non";
    },
    disabledSubmit() {
      return this.answers.length < 2 || this.question === "";
    }
  },
  methods: {
    ...mapActions("app", ["addNotification"]),
    async submitPoll() {
      try {
        await createPoll({
          question: this.question,
          description: this.description,
          answers: this.answers,
          owner: this.userId,
          users: [this.userId],
          isPublic: this.isPublic,
          isActive: true,
          token: randomString({ length: 20 })
        });

        this.addNotification({
          text: "Scrutin créé mon pote",
          status: "success"
        });

        this.question = "";
        this.description = "";
        this.answer = "";
        this.answers = [];
      } catch (error) {
        this.addNotification({
          text: `Il y a eu un problème lors de la création du scrutin`,
          status: "error"
        });
      }
    },
    addAnswer() {
      const cleanAnswer = this.answer.trim();
      if (cleanAnswer.length === 0) {
        this.addNotification({
          text: "Une réponse ne peut pas être vide",
          status: "warning"
        });
        return;
      }

      if (this.answers.find(answer => answer.value === cleanAnswer)) {
        this.addNotification({
          text: "Cette réponse existe déjà",
          status: "warning"
        });

        return;
      }

      this.answers = [
        ...this.answers,
        {
          value: cleanAnswer,
          slug: slugify(cleanAnswer, { lower: true, replacement: "-" })
        }
      ];
      this.answer = "";
      this.$refs.answerInput.focus();
    },
    removeAnswer(slug) {
      this.answers = this.answers.filter(answer => answer.slug !== slug);
    }
  }
};
</script>

<style></style>
