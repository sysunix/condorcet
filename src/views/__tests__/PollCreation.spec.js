import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import PollCreation from "../PollCreation.vue";

import { createPoll } from "../../utils/request";
import { getTranslation } from "../../utils/test";

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock("../../firebase");

jest.mock("../../utils/request");

describe("PollCreation.vue", () => {
  let wrapper;
  let appActions;

  beforeEach(() => {
    jest.clearAllMocks();
    appActions = { addNotification: jest.fn() };

    const store = new Vuex.Store({
      modules: {
        app: {
          namespaced: true,
          actions: appActions,
          state: { featureFlipping: { publicPolls: true } }
        },
        user: {
          namespaced: true
        }
      }
    });

    wrapper = shallowMount(PollCreation, {
      localVue,
      store,
      mocks: {
        $t: getTranslation
      },
      computed: {
        userId: () => "azfnez893jfz"
      }
    });
  });

  it("should render", () => {
    expect(wrapper.vm.userId).toBe("azfnez893jfz");
    expect(wrapper).toMatchSnapshot();
  });

  it("should not add answer to the poll", () => {
    wrapper.vm.addAnswer();

    expect(appActions.addNotification).toHaveBeenCalled();
  });

  it("should remove answer", async () => {
    wrapper.setData({ answers: [{ slug: "foo", value: "Foo" }] });
    expect(wrapper.vm.answers).toEqual([{ slug: "foo", value: "Foo" }]);

    await localVue.nextTick();

    wrapper.find("[data-testid='form-answerProposal-foo']").trigger("click");

    expect(wrapper.vm.answers).toEqual([]);
  });

  it("should add answer to the poll", () => {
    const answerInput = wrapper.find("#answer");
    const addAnwserButton = wrapper.find("button");

    answerInput.setValue("Les pommes");
    answerInput.trigger("change");

    expect(wrapper.vm.answer).toBe("Les pommes");

    addAnwserButton.trigger("click");

    expect(wrapper.vm.answer).toBe("");
    expect(wrapper.vm.answers).toEqual([
      {
        slug: "les-pommes",
        value: "Les pommes"
      }
    ]);
  });

  it("should not create poll and notify user that there is an error if question or answer are not set", () => {
    const questionInput = wrapper.find("#question");
    const answerInput = wrapper.find("#answer");
    const addAnwserButton = wrapper.find("button");

    questionInput.setValue("Quel est le meilleru fruit ?");
    answerInput.trigger("change");

    answerInput.setValue("Les pommes");
    answerInput.trigger("change");
    addAnwserButton.trigger("click");

    answerInput.setValue("Les poires");
    answerInput.trigger("change");
    addAnwserButton.trigger("click");

    wrapper.vm.submitPoll();

    expect(createPoll).toHaveBeenCalled();
  });

  // it("should create the poll and notify user that there is an error", () => {
  //   wrapper.setData({
  //     question: "Qui est la plus belle ?",
  //     answers: ["Megan Fox", "Scarlett", "Natalie Portman"]
  //   });

  //   wrapper.vm.submitPoll();

  //   expect(firebase.db.collection).toHaveBeenCalledWith("polls");
  // });
});
