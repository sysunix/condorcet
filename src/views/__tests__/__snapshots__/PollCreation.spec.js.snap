// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollCreation.vue should render 1`] = `
<form autocomplete="off" class="w-full pt-4">
  <div class="flex flex-wrap -mx-3 mb-6">
    <div class="w-full md:w-1/2 px-3 mb-6 md:mb-0"><label for="question" class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2">Question *</label> <input id="question" data-testid="form-question" type="text" class="appearance-none block w-full bg-gray-200 text-gray-700 border rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"> <label for="description" class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2">Descritpion</label> <textarea data-testid="form-description" class="appearance-none block w-full bg-gray-200 text-gray-700 border rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"></textarea>
      <div class="md:flex md:items-center mb-6"><label for="description" class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2">Rendre le scrutin public ?</label> <input id="isPublic" data-testid="form-isPublic" type="checkbox" name="isPublic" class="mr-2 leading-tight"> <label for="isPublic" class="md:w-2/3 block text-gray-500 font-bold inline-block"><span class="text-sm">Non</span></label></div>
    </div>
    <div class="w-full md:w-1/2 px-3"><label for="answer" class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2">Ajouter une réponse *</label>
      <div class="flex"><input id="answer" data-testid="form-answer" type="text" class="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"> <button data-testid="form-addAnswer" class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4">
          Ajouter
        </button></div>
      <ul class="list-disc mt-6 ml-10"></ul>
    </div>
    <div class="w-full mt-4 px-3 flex justify-center"><button data-testid="form-submit" disabled="disabled" class="w-1/3 bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded opacity-50 cursor-not-allowed">
        Créer
      </button></div>
  </div>
</form>
`;
